src/UserSpaceInstrumentation/InjectLibraryInTracee.cpp

```diff
@@ -0,0 +1,298 @@
+
+// In certain error conditions the tracee is damaged and we don't try to recover from that. We just
+// abort with a fatal log message. None of these errors are expected to occur in operation
+// obvioulsy. That's what the *OrDie methods below are for.
+void FreeMemoryOrDie(pid_t pid, uint64_t address_code, uint64_t size) {
+  auto result = FreeInTracee(pid, address_code, size);
+  FAIL_IF(result.has_error(), "Unable to free previously allocated memory in tracee: \"%s\"",
+          result.error().message());
+}
+
+void RestoreRegistersOrDie(RegisterState& register_state) {
+  auto result = register_state.RestoreRegisters();
+  FAIL_IF(result.has_error(), "Unable to restore register state in tracee: \"%s\"",
+          result.error().message());
+}
+
+uint64_t GetReturnValueOrDie(pid_t pid) {
+  RegisterState return_value_registers;
+  auto result_return_value = return_value_registers.BackupRegisters(pid);
+  FAIL_IF(result_return_value.has_error(), "Unable to read registers after function called :\"%s\"",
+          result_return_value.error().message());
+  return return_value_registers.GetGeneralPurposeRegisters()->x86_64.rax;
+}
+
+// Execute the code at `address_code`. The code segment has to end with an `int3`.
+void ExecuteOrDie(pid_t pid, RegisterState& original_registers, uint64_t address_code) {
+  RegisterState registers_set_rip = original_registers;
+  registers_set_rip.GetGeneralPurposeRegisters()->x86_64.rip = address_code;
+  auto result_restore_registers = registers_set_rip.RestoreRegisters();
+  if (result_restore_registers.has_error()) {
+    FATAL("Unable to continue tracee with PTRACE_CONT.");
+  }
+  if (ptrace(PTRACE_CONT, pid, 0, 0) != 0) {
+    FATAL("Unable to continue tracee with PTRACE_CONT.");
+  }
+  int status = 0;
+  pid_t waited = waitpid(pid, &status, 0);
+  if (waited != pid || !WIFSTOPPED(status) || WSTOPSIG(status) != SIGTRAP) {
+    FATAL("Failed to wait for sigtrap after PTRACE_CONT.");
+  }
+}
+
+}  // namespace
+
+[[nodiscard]] ErrorMessageOr<void*> DlopenInTracee(pid_t pid, std::filesystem::path path,
+                                                   uint32_t flag) {
+  // Figure out address of dlopen in libc.
+  OUTCOME_TRY(address_dlopen, FindFunctionAddress(pid, "dlopen", "libc-"));
+
+  // Backup registers.
+  RegisterState original_registers;
+  OUTCOME_TRY(original_registers.BackupRegisters(pid));
+
+  // Allocate small memory area in the tracee. This is used for the code and the path name.
+  const uint64_t path_length = path.string().length() + 1;  // Include terminating zero.
+  const uint64_t memory_size = kScratchPadSize + path_length;
+  OUTCOME_TRY(address_code, AllocateInTracee(pid, memory_size));
+
+  // Write the name of the .so into memory at address_code with offset of kScratchPadSize.
+  std::vector<uint8_t> path_as_vector(path_length);
+  memcpy(path_as_vector.data(), path.c_str(), path_length);
+  const uint64_t address_so_path = address_code + kScratchPadSize;
```

Comment:
`so_path_address`?

Reply:
Same as above. "Address of .so Path"

Reply:
Same as above, either `address_of_so_path` or `so_path_address`, in English "address so path" reads wrong.

Reply:
see above.